parsers/ruby.c

```diff
@@ -117,31 +112,104 @@ static bool canMatch (const unsigned char** s, const char* literal,
 	return true;
 }
 
+static bool isIdentChar (int c)
+{
+	return (isalnum (c) || c == '_');
+}
+
 static bool notIdentChar (int c)
 {
 	return ! isIdentChar (c);
 }
 
+static bool operatorChar (int c)
+{
+	return (c == '[' || c == ']' ||
+	        c == '=' || c == '!' || c == '~' ||
+	        c == '+' || c == '-' ||
+	        c == '@' || c == '*' || c == '/' || c == '%' ||
+	        c == '<' || c == '>' ||
+	        c == '&' || c == '^' || c == '|');
+}
+
 static bool notOperatorChar (int c)
 {
-	return ! (c == '[' || c == ']' ||
-	          c == '=' || c == '!' || c == '~' ||
-	          c == '+' || c == '-' ||
-	          c == '@' || c == '*' || c == '/' || c == '%' ||
-	          c == '<' || c == '>' ||
-	          c == '&' || c == '^' || c == '|');
+	return ! operatorChar (c);
 }
 
 static bool isWhitespace (int c)
 {
 	return c == 0 || isspace (c);
 }
 
+/*
+ * Advance 's' while the passed predicate is true. Returns true if
+ * advanced by at least one position.
+ */
+static bool advanceWhile (const unsigned char** s, bool (*predicate) (int))
+{
+	const int s_length = strlen ((const char *)*s);
+	unsigned char this_char;
+	int i = 0;
+
+	for (i = 0; i < s_length; ++i)
```

Comment:
This function scans the input twice: once by strlen and once by this `for`.
How about replacing them with following `while`?

```
while (**s != `\0`)
```